collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java

```diff
@@ -0,0 +1,225 @@
+        externalDataReader.doImport(formDefToCsvMedia);
+
+        assertThat(dbFile.exists(), is(true));
+        assertThat(csvFile.exists(), is(true));
+    }
+
+    @Test
+    public void createAndPopulateMetadataTable_createsMetadataTableWithExpectedTimestamp() {
+        final String testMetadataTable = "testMetadataTable";
+
+        SQLiteDatabase.OpenParams.Builder paramsBuilder = new SQLiteDatabase.OpenParams.Builder();
+        SQLiteDatabase db = SQLiteDatabase.createInMemory(paramsBuilder.build());
+        ExternalSQLiteOpenHelper.createAndPopulateMetadataTable(db, testMetadataTable, csvFile);
+
+        assertThat(SQLiteUtils.doesTableExist(db, testMetadataTable), is(true));
+
+        final String[] columnNames = {COLUMN_LAST_MODIFIED};
+        final String selectCriteria = CustomSQLiteQueryBuilder.formatCompareEquals(
+                COLUMN_DATASET_FILENAME,
+                CustomSQLiteQueryBuilder.quoteStringLiteral(SIMPLE_SEARCH_EXTERNAL_CSV_FILENAME));
+        Cursor cursor = db.query(testMetadataTable, columnNames, selectCriteria, null, null, null, null);
+        cursor.moveToFirst();
+        long fileTimestamp = cursor.getLong(0);
+
+        assertThat(fileTimestamp, is(csvFile.lastModified()));
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfDatabaseFileIsDeleted() {
+        // Create the DB file with an initial import
+        ExternalDataReader externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+
+        dbFile.delete();
+
+        // Reimport
+        externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfMetadataTableIsMissing() {
+        // Create the DB file with an initial import
+        ExternalDataReader externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+
+        // Remove the metadata table (mimicking prior versions without the metadata table)
+        SQLiteDatabase db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        SQLiteUtils.dropTable(db, EXTERNAL_METADATA_TABLE_NAME);
+        db.close();
+
+        // Reimport
+        externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+        db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        assertThat("metadata table should be recreated", SQLiteUtils.doesTableExist(db, EXTERNAL_METADATA_TABLE_NAME));
+        db.close();
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfFileIsUpdated() throws IOException, InterruptedException {
```

Comment:
Note that this test has changed. It previously faked an update by changing the db timestamp. Now it actually tests a real file change.